core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -181,7 +181,7 @@ public static Query create(
         result.queryManager.addOutputInfoListener(result.getQueryId(), result::setQueryOutputInfo);
 
         result.queryManager.addStateChangeListener(result.getQueryId(), state -> {
-            if (state.isDone()) {
+            if (!result.exchangeClient.isFinished()) {
```

Comment:
Actually I prefered previous shape with explanation you put in the PR; maybe just convert it to comment in the code.
No we are closing exchange on ANY state change - which is weird. The question which arises why do we need to open it at all in the first place.